kroxylicious/src/test/java/io/kroxylicious/proxy/internal/FilterHandlerTest.java

```diff
@@ -52,6 +63,150 @@ public void testForwardRequest() {
+                        true));
+    }
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource
+    void responseFilterClosesChannel(String name,
+                                     BiFunction<ResponseHeaderData, ApiMessage, CompletableFuture<ResponseFilterResult>> stageFunction,
+                                     boolean forwardExpected) {
+        ApiVersionsResponseFilter filter = (apiVersion, header, response, context) -> stageFunction.apply(header, response);
+        buildChannel(filter);
+        var frame = writeResponse(new ApiVersionsResponseData());
+        channel.runPendingTasks();
+
+        assertThat(channel.isOpen()).isFalse();
+        var propagated = channel.readInbound();
+        if (forwardExpected) {
+            assertThat(propagated).isEqualTo(frame);
+        }
+        else {
+            assertThat(propagated).isNull();
+        }
+    }
+
+    @Test
+    void closedChannelIgnoresDeferredPendingRequests() {
+        var seen = new ArrayList<ApiMessage>();
+        var filterFuture = new CompletableFuture<RequestFilterResult>();
+        ApiVersionsRequestFilter filter = (apiVersion, header, request, context) -> {
+            seen.add(request);
+            return filterFuture;
+        };
+        buildChannel(filter);
+        var frame1 = writeRequest(new ApiVersionsRequestData());
+        writeRequest(new ApiVersionsRequestData().setClientSoftwareName("should not be processed"));
+        // the filter handler will have queued up the second request, awaiting the completion of the first.
+        filterFuture.complete(new RequestFilterResultBuilderImpl().withCloseConnection().build());
+        channel.runPendingTasks();
+
+        assertThat(channel.isOpen()).isFalse();
+        var propagated = channel.readOutbound();
+        assertThat(propagated).isNull();
+        assertThat(seen).hasSize(1);
+        assertThat(seen).containsExactly(frame1.body());
+    }
+
+    @Test
+    void closedChannelIgnoresDeferredPendingResponse() {
+        var seen = new ArrayList<ApiMessage>();
+        var filterFuture = new CompletableFuture<ResponseFilterResult>();
+        ApiVersionsResponseFilter filter = (apiVersion, header, response, context) -> {
+            seen.add(response);
+            return filterFuture;
+        };
+        buildChannel(filter);
+        var frame1 = writeResponse(new ApiVersionsResponseData().setErrorCode((short) 1));
+        writeResponse(new ApiVersionsResponseData().setErrorCode((short) 2));
+        // the filter handler will have queued up the second response, awaiting the completion of the first.
+        filterFuture.complete(new ResponseFilterResultBuilderImpl().withCloseConnection().build());
+        channel.runPendingTasks();
+
+        assertThat(channel.isOpen()).isFalse();
+        var propagated = channel.readInbound();
+        assertThat(propagated).isNull();
+        assertThat(seen).hasSize(1);
+        assertThat(seen).containsExactly(frame1.body());
```

Comment:
One can chain these 
```suggestion
        assertThat(seen).hasSize(1).containsExactly(frame1.body());
```

Though when using `containsExactly` does adding a `size == 1` assertion add anything?